python-frontend/src/main/java/org/sonar/python/types/TypeShed.java

```diff
@@ -66,25 +67,33 @@ private TypeShed() {
       for (Symbol globalVariable : fileInput.globalVariables()) {
         builtins.put(globalVariable.fullyQualifiedName(), globalVariable);
       }
+      TypeShed.builtins = Collections.unmodifiableMap(builtins);
       BaseTreeVisitor visitor = new BaseTreeVisitor() {
         @Override
         public void visitFunctionDef(FunctionDef functionDef) {
-          TypeAnnotation returnTypeAnnotation = functionDef.returnTypeAnnotation();
-          Optional.ofNullable(functionDef.name().symbol()).ifPresent(symbol -> {
-            if (symbol.kind() == Symbol.Kind.FUNCTION && returnTypeAnnotation != null) {
-              FunctionSymbolImpl functionSymbol = (FunctionSymbolImpl) symbol;
-              functionSymbol.setDeclaredReturnType(InferredTypes.declaredType(returnTypeAnnotation, builtins));
-            }
-          });
+          Optional.ofNullable(functionDef.name().symbol()).ifPresent(symbol -> setDeclaredReturnType(symbol, functionDef));
           super.visitFunctionDef(functionDef);
         }
       };
       fileInput.accept(visitor);
-      TypeShed.builtins = Collections.unmodifiableMap(builtins);
     }
     return builtins;
   }
 
+  private static void setDeclaredReturnType(Symbol symbol, FunctionDef functionDef) {
+    TypeAnnotation returnTypeAnnotation = functionDef.returnTypeAnnotation();
+    if (returnTypeAnnotation == null) {
+      return;
+    }
+    if (symbol.is(Symbol.Kind.FUNCTION)) {
+      FunctionSymbolImpl functionSymbol = (FunctionSymbolImpl) symbol;
+      functionSymbol.setDeclaredReturnType(InferredTypes.declaredType(returnTypeAnnotation, builtins));
+    } else if (symbol.is(Symbol.Kind.AMBIGUOUS)) {
+      AmbiguousSymbolImpl ambiguousSymbol = (AmbiguousSymbolImpl) symbol;
+      setDeclaredReturnType(ambiguousSymbol.getSymbol(functionDef.name()), functionDef);
```

Comment:
I think we should find a way to do that differently, without needing `AmbiguousSymbolImpl` to store `symbolsByDeclarationTree`.

Reply:
One idea:
Add a `symbol` method to FunctionDefImpl. This symbol could be different from the symbol attached to the name of the function in the case of an ambiguous symbol.

Reply:
Good idea, I will go for it! Thanks!